### PR TITLE
Refactor `ModuleIO::read/write_cube()`

### PR DESCRIPTION
**Pull request by PeizeLin**
Move some codes in `ModuleIO::read_cube()` and `ModuleIO::write_cube()` to new functions `ModuleIO::read_cube_core_match()`, `ModuleIO::read_cube_core_mismatch()` and `ModuleIO::write_cube_core()`.
Refactor these functions.

For users, there is no change in input or output.

### CODE REVIEW

**Comment by kirk0830** [comment]
@mohanchen this PR is ready for review, except precommit cannnot push format changes.

File: source/module_io/cube_io.h [hunk 1]
```diff
@@ -8,45 +8,91 @@
 
 namespace ModuleIO
 {
-bool read_cube(
+extern bool read_cube(
```
Comment (mohanchen) [inline source/module_io/cube_io.h]: why you add 'extern' before read_cube?

File: source/module_io/cube_io.h [hunk 2]
```diff
@@ -8,45 +8,91 @@
...
+    const UnitCell*const ucell,
+    const int precision = 11,
+    const int out_fermi = 1); // mohan add 2007-10-17
+
+
+// when MPI:
+//      read file as order (ixy,iz) to data[ixy*nz+iz]
+// when serial:
+//      read file as order (ixy,iz) to data[iz*nxy+ixy]
+extern void read_cube_core_match(
```
Comment (kirk0830) [inline source/module_io/cube_io.h]: could you add some annotation to describe what this function for? I can hardly understand why reading the cube file depends on whether the core match or mismatch...

File: source/module_io/cube_io.h [hunk 3]
```diff
@@ -78,15 +124,15 @@ void write_cube(
      * @param nz the dimension of grids along z
      * @param data the interpolated results
      */
-    void trilinear_interpolate(std::ifstream& ifs,
+    extern void trilinear_interpolate(std::ifstream& ifs,
```
Comment (kirk0830) [inline source/module_io/cube_io.h]: as @mohanchen has commented and what I commented before, the use of extern seems not necessary.

File: source/module_io/write_cube.cpp [hunk 4]
```diff
@@ -127,8 +127,43 @@ void ModuleIO::write_cube(
...
+
+        /// for cube file
+        ofs_cube.close();
+    }
+
+    return;
+}
+
+
+void ModuleIO::write_cube_core(
```
Comment (kirk0830) [inline source/module_io/write_cube.cpp:149]: From your PR I can realize that maybe the original implementation of writing or even together with reading the cube file really is/are bad, it is the reason why you have to implement different functions to write cube file, even if the way to organize the cube file is almost unique...what should do is to reorganize the data to fit the structure that required by the cube file before calling a function to really write the cube file, I think.

**Review by kirk0830 — CHANGES_REQUESTED** [review]

**Review by kirk0830 — COMMENTED** [review]
This PR refactors partially the cube io functions, separating the trilinear interpolation from the kernel function of read and write cube file. Additionally, the parameters lists are slightly modified, mainly in two aspects, drop the "&" behind "const int" and add const to pointers to array. 
There should still be further refactor because there is still not a pure, clean function to do the only one thing that read or write cube file, hope some developer will so this later.
The PR seems to be the prerequisite of DFPT development for outputting Vxc on realspace grid.

File: source/module_io/cube_io.h (hunk 1 above)
Comment (kirk0830) [reply source/module_io/cube_io.h]: @mohanchen the extern prefix has been deleted

**Review by mohanchen — APPROVED** [review]

**Review by kirk0830 — APPROVED** [review]
this PR can be merged with an issue submitted #5191.

**Pull request merged**